shared_model/backend/protobuf/common_objects/proto_common_objects_factory.hpp

```diff
@@ -0,0 +1,259 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_PROTO_COMMON_OBJECTS_FACTORY_HPP
+#define IROHA_PROTO_COMMON_OBJECTS_FACTORY_HPP
+
+#include <regex>
+
+#include "backend/protobuf/common_objects/account.hpp"
+#include "backend/protobuf/common_objects/account_asset.hpp"
+#include "backend/protobuf/common_objects/asset.hpp"
+#include "backend/protobuf/common_objects/domain.hpp"
+#include "backend/protobuf/common_objects/peer.hpp"
+#include "backend/protobuf/common_objects/signature.hpp"
+#include "common/result.hpp"
+#include "interfaces/common_objects/common_objects_factory.hpp"
+#include "primitive.pb.h"
+#include "validators/answer.hpp"
+
+namespace shared_model {
+  namespace proto {
+
+  /**
+   * ProtoCommonObjectsFactory constructs protobuf-based objects.
+   * It performs stateful validation with provided validator
+   * @tparam Validator
+   */
+    template <typename Validator>
+    class ProtoCommonObjectsFactory : public interface::CommonObjectsFactory {
+     public:
+      FactoryResult<std::unique_ptr<interface::Peer>> createPeer(
+          const interface::types::AddressType &address,
+          const interface::types::PubkeyType &public_key) override {
+        iroha::protocol::Peer peer;
+        peer.set_address(address);
+        peer.set_peer_key(crypto::toBinaryString(public_key));
+        auto proto_peer = std::make_unique<Peer>(std::move(peer));
+
+        auto answer =
+            validate(*proto_peer, [this](const auto &peer, auto &reasons) {
+              validator_.validatePeer(reasons, peer);
+            });
+
+        if (answer) {
```

Comment:
Doesn't look very readable. Maybe, you should rename the variable to ```errors```? It represents something wrong happened only, so I think it's a reasonable name

Reply:
If you choose to follow the advice, change names of all variables in below functions as well

Reply:
Answer can be empty, which indicates lack of errors. But i agree that errors is a better name here